File: backend/tags.py
"""Tag parsing utilities for todo text.

Tags are written inline as #tagname (e.g. "Fix login bug #backend #urgent").
They are stored as part of the todo text — no separate model field needed.

Priority keywords (#p1–#p4 and aliases like #critical, #high, #medium, #low)
are parsed as priorities, NOT as regular tags.
"""

from __future__ import annotations

import re
from typing import Optional

# Matches #tag where tag is alphanumeric (plus hyphens/underscores), at least 1 char.
# Must be preceded by whitespace or start-of-string to avoid matching e.g. "C#" or URLs with #fragments.
TAG_RE = re.compile(r"(?:^|(?<=\s))#([A-Za-z][A-Za-z0-9_-]*)")

# Priority keyword mapping: keyword -> priority level (1=critical, 4=low)
PRIORITY_KEYWORDS: dict[str, int] = {
    "p1": 1, "critical": 1,
    "p2": 2, "high": 2,
    "p3": 3, "medium": 3,
    "p4": 4, "low": 4,
}

# Reverse: priority level -> canonical short label
PRIORITY_LABELS: dict[int, str] = {1: "p1", 2: "p2", 3: "p3", 4: "p4"}


def is_priority_keyword(tag: str) -> bool:
    """Check if a tag name (without #) is a priority keyword."""
    return tag.lower() in PRIORITY_KEYWORDS


def parse_priority(text: str) -> Optional[int]:
    """Extract the highest priority level from text (lowest number wins). Returns None if no priority keyword found."""
    best: Optional[int] = None
    for m in TAG_RE.finditer(text):
        tag = m.group(1).lower()
        if tag in PRIORITY_KEYWORDS:
            level = PRIORITY_KEYWORDS[tag]
            if best is None or level < best:
                best = level
    return best


def parse_tags(text: str) -> list[str]:
    """Extract tags from text, returning lowercase deduplicated list. Excludes priority keywords."""
    seen: set[str] = set()
    result: list[str] = []
    for m in TAG_RE.finditer(text):
        tag = m.group(1).lower()
        if tag not in seen and not is_priority_keyword(tag):
            seen.add(tag)
            result.append(tag)
    return result


def strip_tags_from_text(text: str) -> str:
    """Remove #tag tokens from text for display purposes."""
    return TAG_RE.sub("", text).strip()


def strip_priority_from_text(text: str) -> str:
    """Remove priority keyword tokens from text for display purposes."""
    def _replace(m: re.Match) -> str:
        if is_priority_keyword(m.group(1)):
            return ""
        return m.group(0)
    return TAG_RE.sub(_replace, text).strip()


def filter_unknown_tags(text: str, known_tags: set[str]) -> str:
    """Remove tags from text that aren't in the known set. Used for Claude-created todos."""
    def _replace(m: re.Match) -> str:
        tag = m.group(1).lower()
        if tag in known_tags:
            return m.group(0)  # keep it
        return ""  # strip unknown tag
    return TAG_RE.sub(_replace, text).strip()


def rename_tag_in_text(text: str, old_tag: str, new_tag: str) -> str:
    """Replace occurrences of #old_tag with #new_tag in text (case-insensitive match)."""
    def _replace(m: re.Match) -> str:
        if m.group(1).lower() == old_tag.lower():
            return m.group(0)[:len(m.group(0)) - len(m.group(1))] + new_tag
        return m.group(0)
    return TAG_RE.sub(_replace, text)


def collect_all_tags(todo_texts: list[str]) -> list[str]:
    """Collect all unique tags across multiple todo texts, sorted alphabetically."""
    all_tags: set[str] = set()
    for text in todo_texts:
        all_tags.update(parse_tags(text))
    return sorted(all_tags)

File: frontend/src/utils/tags.ts
/**
 * Tag parsing utilities — mirrors backend/tags.py logic.
 * Tags are #word tokens in todo text (e.g. "Fix bug #backend #urgent").
 * Priority keywords (#p1–#p4, #critical, #high, #medium, #low) are parsed
 * as priorities, NOT as regular tags.
 */

const TAG_RE = /(?:^|(?<=\s))#([A-Za-z][A-Za-z0-9_-]*)/g;

/** Priority keyword mapping: keyword -> priority level (1=critical, 4=low) */
export const PRIORITY_KEYWORDS: Record<string, number> = {
  p1: 1, critical: 1,
  p2: 2, high: 2,
  p3: 3, medium: 3,
  p4: 4, low: 4,
};

/** Priority level -> display info */
export const PRIORITY_INFO: Record<number, { label: string; short: string; color: string }> = {
  1: { label: "Critical", short: "P1", color: "#e53e3e" },
  2: { label: "High", short: "P2", color: "#dd6b20" },
  3: { label: "Medium", short: "P3", color: "#d69e2e" },
  4: { label: "Low", short: "P4", color: "#718096" },
};

/** Check if a tag name (without #) is a priority keyword. */
export function isPriorityKeyword(tag: string): boolean {
  return tag.toLowerCase() in PRIORITY_KEYWORDS;
}

/** Extract the highest priority from text. Returns null if none found. */
export function parsePriority(text: string): number | null {
  let best: number | null = null;
  for (const m of text.matchAll(TAG_RE)) {
    const tag = m[1].toLowerCase();
    const level = PRIORITY_KEYWORDS[tag];
    if (level !== undefined && (best === null || level < best)) {
      best = level;
    }
  }
  return best;
}

/** Extract tags from text, returning lowercase deduplicated list. Excludes priority keywords. */
export function parseTags(text: string): string[] {
  const seen = new Set<string>();
  const result: string[] = [];
  for (const m of text.matchAll(TAG_RE)) {
    const tag = m[1].toLowerCase();
    if (!seen.has(tag) && !(tag in PRIORITY_KEYWORDS)) {
      seen.add(tag);
      result.push(tag);
    }
  }
  return result;
}

/** Remove #tag tokens from text for display purposes. */
export function stripTagsFromText(text: string): string {
  return text.replace(TAG_RE, "").replace(/\s{2,}/g, " ").trim();
}

/** Remove priority keyword tokens from text for display purposes. */
export function stripPriorityFromText(text: string): string {
  return text.replace(TAG_RE, (match, tag: string) => {
    return tag.toLowerCase() in PRIORITY_KEYWORDS ? "" : match;
  }).replace(/\s{2,}/g, " ").trim();
}
